llpc/lower/llpcSpirvLowerAccessChain.cpp

```diff
@@ -63,12 +63,69 @@ bool SpirvLowerAccessChain::runImpl(Module &module) {
 
   SpirvLower::init(&module);
 
-  // Invoke handling of "getelementptr" instruction
+  // Invoke handling of "getelementptr", "load" and "store" instructions
   visit(m_module);
 
   return true;
 }
 
+// =====================================================================================================================
+// Checks if pointer operand of getelementptr is a global value and if types are the same.
+// If types are different (which may happen for opaque pointers) then this function is adding the missing zero-index
+// elements to the gep instruction.
+//
+// One of the examples may be a type in which we have a multiple nested structures.
+// { { [4 x float] } }
+//
+// @param gep : Getelementptr instruction.
+void SpirvLowerAccessChain::tryToAddMissingIndicesBetweenGVandGEP(GEPOperator *gep) {
+
+  // We are interested only in address spaces which are used while doing global value lowering for store and load.
+  const unsigned addrSpace = gep->getType()->getPointerAddressSpace();
+  if (addrSpace != SPIRAS_Input && addrSpace != SPIRAS_Output && addrSpace != SPIRAS_TaskPayload)
+    return;
+
+  GlobalValue *gv = dyn_cast<GlobalValue>(gep->getPointerOperand());
+  if (!gv)
+    return;
+
+  // No missing indices, types are the same.
+  if (gep->getSourceElementType() == gv->getValueType())
+    return;
+
+  SmallVector<Value *, 8> idxs;
+  idxs.push_back(m_builder->getInt32(0));
+  appendZeroIndexToMatchTypes(idxs, gep->getSourceElementType(), gv->getValueType());
+
+  for (unsigned i = 2; i != gep->getNumOperands(); ++i)
+    idxs.push_back(gep->getOperand(i));
+
+  Value *newGep = m_builder->CreateGEP(gv->getValueType(), gv, idxs);
+  gep->replaceAllUsesWith(newGep);
+  if (Instruction *inst = dyn_cast<Instruction>(gep))
+    inst->eraseFromParent();
+}
+
+// =====================================================================================================================
+// Visits "load" instruction
+//
+// @param loadInst : "Load" instruction
+void SpirvLowerAccessChain::visitLoadInst(LoadInst &loadInst) {
+  if (GEPOperator *gep = dyn_cast<GEPOperator>(loadInst.getPointerOperand())) {
+    tryToAddMissingIndicesBetweenGVandGEP(gep);
+  }
+}
+
+// =====================================================================================================================
+// Visits "store" instruction
+//
+// @param storeInst : "Store" instruction
+void SpirvLowerAccessChain::visitStoreInst(StoreInst &storeInst) {
+  if (GEPOperator *gep = dyn_cast<GEPOperator>(storeInst.getPointerOperand())) {
+    tryToAddMissingIndicesBetweenGVandGEP(gep);
+  }
+}
```

Comment:
These functions need to set the builder insert point, don't they?

Reply:
You are referring to `getBuilder()->SetInsertPoint` ?
But do I need to set inset point if I'm creating new instruction (in this case new GEP) and replacing it with existing GEP ?
I see that `tryToCoalesceChain` is doing more or less the same (creating new single GEP and replacing old one) and it is not using `SetInsertPoint`.